Source/Urho3D/Input/DirectionalPadAdapter.cpp

```diff
@@ -0,0 +1,405 @@
+void DirectionalPadAdapter::SendKeyDown(Scancode key)
+{
+    using namespace KeyDown;
+
+    VariantMap args;
+
+    switch (key)
+    {
+    case SCANCODE_UP: args[P_KEY] = KEY_UP; break;
+    case SCANCODE_DOWN: args[P_KEY] = KEY_DOWN; break;
+    case SCANCODE_LEFT: args[P_KEY] = KEY_LEFT; break;
+    case SCANCODE_RIGHT: args[P_KEY] = KEY_RIGHT; break;
+    default: args[P_KEY] = KEY_UNKNOWN; break;
+    }
+
+    args[P_SCANCODE] = key;
+    args[P_BUTTONS] = 0;
+    args[P_QUALIFIERS] = 0;
+    args[P_REPEAT] = false;
+    SendEvent(E_KEYDOWN, args, false);
+}
+
+void DirectionalPadAdapter::SendKeyUp(Scancode key)
+{
+    using namespace KeyUp;
+
+    VariantMap args;
+
+    using namespace KeyUp;
+    switch (key)
+    {
+    case SCANCODE_UP: args[P_KEY] = KEY_UP; break;
+    case SCANCODE_DOWN: args[P_KEY] = KEY_DOWN; break;
+    case SCANCODE_LEFT: args[P_KEY] = KEY_LEFT; break;
+    case SCANCODE_RIGHT: args[P_KEY] = KEY_RIGHT; break;
+    default: args[P_KEY] = KEY_UNKNOWN; break;
+    }
+    args[P_SCANCODE] = key;
+    args[P_BUTTONS] = 0;
+    args[P_QUALIFIERS] = 0;
+    SendEvent(E_KEYUP, args, false);
+}
+
+void DirectionalPadAdapter::HandleKeyDown(StringHash eventType, VariantMap& args)
+{
+    using namespace KeyDown;
+    switch (args[P_SCANCODE].GetUInt())
+    {
+    case SCANCODE_W:
+    case SCANCODE_UP:
+        Append(up_, InputType::Keyboard, SCANCODE_UP);
+        break;
+    case SCANCODE_S:
+    case SCANCODE_DOWN:
+        Append(down_, InputType::Keyboard, SCANCODE_DOWN);
+        break;
+    case SCANCODE_A:
+    case SCANCODE_LEFT:
+        Append(left_, InputType::Keyboard, SCANCODE_LEFT);
+        break;
+    case SCANCODE_D:
+    case SCANCODE_RIGHT:
+        Append(right_, InputType::Keyboard, SCANCODE_RIGHT);
+        break;
+    }
+}
+
+void DirectionalPadAdapter::HandleKeyUp(StringHash eventType, VariantMap& args)
+{
+    using namespace KeyUp;
+    switch (args[P_SCANCODE].GetUInt())
+    {
+    case SCANCODE_W:
+    case SCANCODE_UP:
+        Remove(up_, InputType::Keyboard, SCANCODE_UP);
+        break;
+    case SCANCODE_S:
+    case SCANCODE_DOWN:
+        Remove(down_, InputType::Keyboard, SCANCODE_DOWN);
+        break;
+    case SCANCODE_A:
+    case SCANCODE_LEFT:
+        Remove(left_, InputType::Keyboard, SCANCODE_LEFT);
+        break;
+    case SCANCODE_D:
+    case SCANCODE_RIGHT:
+        Remove(right_, InputType::Keyboard, SCANCODE_RIGHT);
+        break;
+    }
+}
+
+void DirectionalPadAdapter::HandleJoystickAxisMove(StringHash eventType, VariantMap& args)
+{
+    using namespace JoystickAxisMove;
+
+
+    auto joystickId = args[P_JOYSTICKID].GetInt();
+    if (joystickId == ignoreJoystickId_)
+        return;
+    
+    auto eventId = static_cast<InputType>(static_cast<unsigned>(InputType::JoystickAxis) + joystickId);
+
+    auto axisIndex = args[P_AXIS].GetUInt();
+    auto value = args[P_POSITION].GetFloat();
+
+    // Up-Down
+    if (axisIndex == 1)
+    {
+        if (value > 0.6f)
```

Comment:
Is it worth being a property?

Reply:
And also axis index should be configurable